src/rust/engine/fs/store/src/lib.rs

```diff
@@ -545,16 +545,23 @@ impl Store {
             let remote = remote.clone();
 
             async move {
-              let executor = local.executor().clone();
-              let maybe_upload = local
+              // We need to copy the bytes into memory so that they may be used safely in an async
+              // future. While this unfortunately increases memory consumption, we prioritize
```

Comment:
... but that's also just being literal with existing APIs / trying to make a minimal change IIUC. The remote store could also have a local store (or get handed one) and be given an entry_type and digest and told to store bytes. It could then handle the shuffle instead of buffer all in its own task.